Index: pkg-tigervnc/CMakeLists.txt
===================================================================
--- pkg-tigervnc.orig/CMakeLists.txt
+++ pkg-tigervnc/CMakeLists.txt
@@ -241,12 +241,18 @@ if(UNIX AND NOT APPLE)
   if(X11_Xinerama_FOUND)
     set(FLTK_LIBRARIES ${FLTK_LIBRARIES} ${X11_Xinerama_LIB})
   endif()
+  if(X11_Xext_FOUND)
+    set(FLTK_LIBRARIES ${FLTK_LIBRARIES} ${X11_Xext_LIB})
+  endif()
   if(X11_Xfixes_FOUND)
     set(FLTK_LIBRARIES ${FLTK_LIBRARIES} ${X11_Xfixes_LIB})
   endif()
   if(X11_Xcursor_FOUND)
     set(FLTK_LIBRARIES ${FLTK_LIBRARIES} ${X11_Xcursor_LIB})
   endif()
+  if(X11_X11_FOUND)
+    set(FLTK_LIBRARIES ${FLTK_LIBRARIES} ${X11_X11_LIB})
+  endif()
   if(X11_Xrender_FOUND)
     set(FLTK_LIBRARIES ${FLTK_LIBRARIES} ${X11_Xrender_LIB})
   endif()
Index: pkg-tigervnc/unix/vncconfig/CMakeLists.txt
===================================================================
--- pkg-tigervnc.orig/unix/vncconfig/CMakeLists.txt
+++ pkg-tigervnc/unix/vncconfig/CMakeLists.txt
@@ -9,7 +9,7 @@ add_executable(vncconfig
   vncconfig.cxx
   QueryConnectDialog.cxx)
 
-target_link_libraries(vncconfig tx rfb network rdr ${X11_LIBRARIES})
+target_link_libraries(vncconfig tx rfb network rdr ${X11_LIBRARIES} ${X11_X11_LIB})
 
 install(TARGETS vncconfig DESTINATION ${BIN_DIR})
 install(FILES vncconfig.man DESTINATION ${MAN_DIR}/man1 RENAME vncconfig.1)
Index: pkg-tigervnc/vncviewer/CMakeLists.txt
===================================================================
--- pkg-tigervnc.orig/vncviewer/CMakeLists.txt
+++ pkg-tigervnc/vncviewer/CMakeLists.txt
@@ -46,7 +46,7 @@ else()
   add_executable(vncviewer ${VNCVIEWER_SOURCES})
 endif()
 
-target_link_libraries(vncviewer rfb network rdr os ${FLTK_LIBRARIES} ${GETTEXT_LIBRARIES})
+target_link_libraries(vncviewer rfb network rdr os ${FLTK_LIBRARIES} ${GETTEXT_LIBRARIES} -lX11)
 
 if(APPLE)
   target_link_libraries(vncviewer "-framework Cocoa" "-framework Carbon")
