--- tigervnc/unix/xserver/os/osinit.c	2012-07-04 01:06:02.000000000 +0200
+++ tigervnc/unix/xserver/os/osinit.c	2012-07-04 01:06:27.000000000 +0200
@@ -140,7 +140,7 @@
 
 #ifdef SA_SIGINFO
   if (sip->si_code == SI_USER) {
-      ErrorF("Recieved signal %d sent by process %ld, uid %ld\n",
+      ErrorF("Received signal %d sent by process %ld, uid %ld\n",
 	     signo, (long) sip->si_pid, (long) sip->si_uid);
   } else {
       switch (signo) {
