Index: b/CMakeLists.txt
===================================================================
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -162,7 +162,11 @@ endif()
 find_package(ZLIB)
 option(USE_INCLUDED_ZLIB "Force use of the bundled zlib")
 if(NOT ZLIB_FOUND)
-  set(USE_INCLUDED_ZLIB 1)
+  if (NOT DEFINED USE_INCLUDED_ZLIB)
+    set(USE_INCLUDED_ZLIB 1)
+  elseif(NOT USE_INCLUDED_ZLIB)
+    message(FATAL_ERROR "System provided zlib library not found and usage of bundled version was disabled")
+  endif()
 endif()
 if(USE_INCLUDED_ZLIB)
   message(STATUS "Using included zlib library")
