--- tigervnc/CMakeLists.txt.orig	2012-06-29 18:32:13.000000000 +0200
+++ tigervnc/CMakeLists.txt	2012-06-29 18:37:07.000000000 +0200
@@ -151,7 +151,11 @@
 find_package(ZLIB)
 option(USE_INCLUDED_ZLIB "Force use of the bundled zlib")
 if(NOT ZLIB_FOUND)
-  set(USE_INCLUDED_ZLIB 1)
+  if (NOT DEFINED USE_INCLUDED_ZLIB)
+    set(USE_INCLUDED_ZLIB 1)
+  elseif(NOT USE_INCLUDED_ZLIB)
+    message(FATAL_ERROR "System provided zlib library not found and usage of bundled version was disabled")
+  endif()
 endif()
 if(USE_INCLUDED_ZLIB)
   message(STATUS "Using included zlib library")
@@ -287,7 +291,11 @@
 if(NOT FLTK_FOUND OR NOT HAVE_FLTK_DEAD_KEYS OR NOT HAVE_FLTK_CLIPBOARD
   OR NOT HAVE_FLTK_MEDIAKEYS OR NOT HAVE_FLTK_FULLSCREEN
   OR NOT HAVE_FLTK_CURSOR)
-  set(USE_INCLUDED_FLTK 1)
+  if (NOT DEFINED USE_INCLUDED_FLTK)
+    set(USE_INCLUDED_FLTK 1)
+  elseif(NOT USE_INCLUDED_FLTK)
+    message(FATAL_ERROR "System provided FLTK library not found and usage of bundled version was disabled")
+  endif()
 endif()
 if(USE_INCLUDED_FLTK)
   # Check that we have the proper dependencies
