--- tigervnc/CMakeLists.txt.orig	2012-07-06 22:14:47.000000000 +0200
+++ tigervnc/CMakeLists.txt	2012-07-06 22:26:17.000000000 +0200
@@ -162,8 +162,8 @@
 endif()
 
 # Check for gettext
-option(ENABLE_NLS "Enable translation of program messages" ON)
-if(ENABLE_NLS)
+option(ENABLE_NLS "Enable translation of program messages")
+if(NOT DEFINED ENABLE_NLS OR ENABLE_NLS)
   # Tools
   find_package(Gettext)
   set(LOCALE_DIR "${CMAKE_INSTALL_PREFIX}/share/locale")
@@ -194,8 +194,14 @@
   endif()
 
   if(NOT GETTEXT_FOUND OR NOT ICONV_FOUND)
-    message(WARNING "Gettext NOT found.  Native Language Support disabled.")
-    set(ENABLE_NLS 0)
+    if(DEFINED ENABLE_NLS)
+      message(FATAL_ERROR "Native Language Support requested, but gettext NOT found!")
+    else()
+      message(WARNING "Gettext NOT found.  Native Language Support disabled.")
+      set(ENABLE_NLS 0)
+    endif()
+  else()
+    set(ENABLE_NLS 1)
   endif()
 endif()
 
@@ -360,8 +366,8 @@
 endif()
 
 # Check for GNUTLS library
-option(ENABLE_GNUTLS "Enable protocol encryption and advanced authentication" ON)
-if(ENABLE_GNUTLS)
+option(ENABLE_GNUTLS "Enable protocol encryption and advanced authentication")
+if(NOT DEFINED ENABLE_GNUTLS OR ENABLE_GNUTLS)
   find_package(GnuTLS)
   if (GNUTLS_FOUND)
     include_directories(${GNUTLS_INCLUDE_DIR})
@@ -391,12 +397,18 @@
     set(CMAKE_REQUIRED_FLAGS)
     set(CMAKE_EXTRA_INCLUDE_FILES) 
     set(CMAKE_REQUIRED_LIBRARIES)
+  else()
+    if(DEFINED ENABLE_GNUTLS)
+      message(FATAL_ERROR "GnuTLS support requested, but library could not be found!")
+    else()
+      message(WARNING "GnuTLS NOT found. SSL support disabled.")
+    endif()
   endif()
 endif()
 
 # Check for PAM library
-option(ENABLE_PAM "Enable PAM authentication support" ON)
-if(ENABLE_PAM)
+option(ENABLE_PAM "Enable PAM authentication support")
+if(NOT DEFINED ENABLE_PAM OR ENABLE_PAM)
   check_include_files(security/pam_appl.h HAVE_PAM_H)
   set(CMAKE_REQUIRED_LIBRARIES -lpam)
   check_function_exists(pam_start HAVE_PAM_START)
@@ -404,7 +416,12 @@
   if(HAVE_PAM_H AND HAVE_PAM_START)
     set(PAM_LIBS pam)
   else()
-    set(ENABLE_PAM 0)
+    if(DEFINED ENABLE_PAM)
+      message(FATAL_ERROR "PAM support requested, but libpam could not be found!")
+    else()
+      message(WARNING "PAM NOT found.")
+      set(ENABLE_PAM 0)
+    endif()
   endif()
 endif()
 set(HAVE_PAM ${ENABLE_PAM})
